packages/react-hook-form/hooks/useFormApollo.tsx

```diff
@@ -0,0 +1,168 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { ApolloError, MutationOptions, QueryOptions, useApolloClient } from '@apollo/client'
+import { useMotionValueValue } from '@graphcommerce/framer-utils'
+import { isMotionValue, MotionValue, useMotionValue } from 'framer-motion'
+import {
+  FieldValues,
+  FormProvider,
+  FormProviderProps,
+  SubmitErrorHandler,
+  useForm,
+  useFormContext,
+  UseFormProps,
+  UseFormReturn,
+} from 'react-hook-form'
+
+export type SubmitValidHandler<Q extends Record<string, unknown>, V extends FieldValues> = (
+  execute: (variables?: V) => Promise<Q>,
+  variables: V,
+) => any | Promise<any>
+
+type HandleSubmit<Q extends Record<string, unknown>, V extends FieldValues> = (
+  onValid?: SubmitValidHandler<Q, V>,
+  onInvalid?: SubmitErrorHandler<V>,
+) => (e?: React.BaseSyntheticEvent) => Promise<void>
+
+export type UseFormGqlMutationReturn<
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Q extends Record<string, unknown> = Record<string, unknown>,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  V extends FieldValues = FieldValues,
+> = Omit<UseFormReturn<V>, 'handleSubmit' | 'formState'> & {
+  data: MotionValue<Q | null | undefined>
+  error: MotionValue<ApolloError | null | undefined>
+  handleSubmit: HandleSubmit<Q, V>
+}
+
+function isMutation<Q, V>(
+  operation: MutationOptions<Q, V> | QueryOptions<V, Q>,
+): operation is MutationOptions<Q, V> {
+  return (operation as MutationOptions<Q, V>).mutation !== undefined
+}
+
+/**
+ * `useApolloForm`
+ *
+ * Example:
+ *
+ * Without any additional logic
+ *
+ * ```tsx
+ * const form = useFormApollo({
+ *   mutation: TestShippingAddressFormDocument,
+ * })
+ *
+ * const onSubmit = form.handleSubmit()
+ * ```
+ *
+ * With additional logic
+ *
+ * ```tsx
+ * const form = useFormApollo({
+ *   mutation: TestShippingAddressFormDocument,
+ * })
+ *
+ * const onSubmit = form.handleSubmit(async (execute, variables) => {
+ *   // Modify variables before executing
+ *   const result = await execute({ ...variables, customerNote: 'joi' })
+ *
+ *   if (!result.errors) {
+ *     // Do something with a successful operation
+ *   }
+ *
+ *   return result
+ * })
+ * ```
+ */
+export function useFormApollo<Q extends Record<string, unknown>, V extends FieldValues>(
+  operation: QueryOptions<V, Q>,
+  useFormProps?: UseFormProps<V>,
+  resetOnResult?: (result: Q) => V,
+): UseFormGqlMutationReturn<Q, V>
+export function useFormApollo<Q extends Record<string, unknown>, V extends FieldValues>(
+  operation: MutationOptions<Q, V>,
+  useFormProps?: UseFormProps<V>,
+  resetOnResult?: (result: Q) => V,
+): UseFormGqlMutationReturn<Q, V>
+export function useFormApollo<Q extends Record<string, unknown>, V extends FieldValues>(
+  operation: MutationOptions<Q, V> | QueryOptions<Partial<V>, Q>,
```

Comment:
- How can we support switching between multiple nutations based on external query data?
- How can we support switching between multiple mutations based on the submit button's value?
- How can we support switching between multiple mutations based on the form data?
- How can we support multiple serial mutations?
- How can we support multiple optional mutations?
- How can we support parallel mutations?
- How can we chain multiple mutations?
- How can we dedupe operations when they don't need to be executed (again)?
- How can we skip operations when they don't need to be executed because they are optional?
- How can we maintain a 'progress' on the submissions?

---

```tsx
const switchOnExternal = useFormApollo(operation: external ? One : Two)
...todo
```